### PR TITLE
feat: add delly CNV, vcf2cytosure

### PR DESCRIPTION
**Pull request by khurrammaqbool**
### This PR:

Added: Delly CNV for tumor-normal analysis
Added: TIDDIT SV for tumor only and tumor-normal
Added: A python script to process CNVs from ascat and delly
Added: vcf2cytosure for tumor only and tumor-normal

For details see: [WGS to cytosure from balsamic?](https://github.com/Clinical-Genomics/BALSAMIC/issues/605)
### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by ashwini06** [comment]
@khurrammaqbool : Can you please edit the PR description or link to the relevant issue to understand the scope of this PR. As this PR includes many file changes, if possible, please detail the major changes done in this PR for a better understanding

File: BALSAMIC/containers/varcall_py3/varcall_py3.yaml [hunk 1]
```diff
@@ -3,12 +3,16 @@ channels:
...
+  - anaconda::click=8.0.4
   - bioconda::bcftools=1.11
   - bioconda::tabix=0.2.6
   - bioconda::samtools=1.11
   - bioconda::gatk=3.8
   - bioconda::vardict=2019.06.04
   - bioconda::vardict-java=1.8.3
   - bioconda::svdb=2.6.0
+  - bioconda::vcfpy=0.13.4
+  - bioconda::bwa=0.7.17
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/containers/varcall_py3/varcall_py3.yaml:15]: It is a dependency.

**Comment by khurrammaqbool** [comment]
> @khurrammaqbool : Can you please edit the PR description or link to the relevant issue to understand the scope of this PR. As this PR includes many file changes, if possible, please detail the major changes done in this PR for a better understanding

The description is updated.

File: BALSAMIC/containers/varcall_py3/Dockerfile [hunk 2]
```diff
@@ -20,4 +20,4 @@ COPY . /${WORK_DIR}
 RUN cd /${WORK_DIR}/BALSAMIC/containers/${CONTAINER_NAME}/ && /bin/sh ${CONTAINER_NAME}.sh ${CONTAINER_NAME}
 
 # Clean work environment
-RUN rm -rf /${WORK_DIR} && conda clean --all --yes
+RUN rm -rf /${WORK_DIR:?} && conda clean --all --yes
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/containers/varcall_py3/Dockerfile:23]: I think it is better to keep :?

File: BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule [hunk 3]
```diff
@@ -1,24 +1,75 @@
...
+            Path(singularity_image, config["bioinfo_tools"].get("vcf2cytosure") + ".sif").as_posix()
+        threads:
+            get_threads(cluster_config, "vcf2cytosure_convert")
+        params:
+            case_name = config["analysis"]["case_id"],
+            housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "cnv-somatic"},
+        message: "Converting CNVs from VCF to the CGH format using vcf2cytosure for {params.case_name}"
+        shell:
+            """
+    vcf2cytosure --vcf {input.delly_vcf} --coverage {input.tiddit_cov_tumor} --out {output.cgh_tumor} --bins 20
```
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule:45]: Adding coverage information from tiddit will require minimum changes in vcf2cytosure.

**Review by ivadym — COMMENTED** [review]
🥇

File: BALSAMIC/assets/scripts/process_CNV.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+)
+@click.option(
+    "-n",
+    "--normal-sample",
+    default="NORMAL",
+    required=False,
+    help="Normal sample ID from vcf header",
+)
+def main(vcf_file, cnv_caller, tumor_sample, normal_sample):
+    vcf = vcfpy.Reader.from_path(vcf_file)
```
Comment (ashwini06) [inline BALSAMIC/assets/scripts/process_CNV.py:54]: Would be nice with the Docstring comments for all the functions in this script

File: BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashwini06) [reply BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule:45]: As different tools use different genomic bin sizes and normalization methods for coverage calculation and the same files is being used for variant calling. Combining coverage and vcf files from different callers, will there be any affect on getting different information in cgh files?

File: BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule [hunk 5]
```diff
@@ -1,24 +1,75 @@
...
+        input:
+            ascat_vcf = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.vcf.gz",
+            tiddit_cov_tumor = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".tumor.tiddit_cov.bed",
+            tiddit_cov_normal = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".normal.tiddit_cov.bed",
+        output:
+            ascat_vcf = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".filtered.ascat.vcf.gz",
+            cgh_tumor = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".tumor.vcf2cytosure.cgh",
+            cgh_normal = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".normal.vcf2cytosure.cgh",
+        benchmark:
+            Path(benchmark_dir, 'vcf2cytosure_convert.' + config["analysis"]["case_id"] + ".tsv").as_posix()
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule:59]:
```suggestion
            Path(benchmark_dir, 'vcf2cytosure_convert_tumor_normal_' + config["analysis"]["case_id"] + ".tsv").as_posix()
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 6]
```diff
@@ -90,26 +90,50 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+        bamT = bam_dir + tumor_bam,
+        map = config["reference"]["delly_mappability"],
     output:
-        vcf = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".dellysv.vcf.gz",
+        cnv_delly = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".delly.bcf",
+        rd_delly = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.cov.gz",
+        namemap= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.sample_name_map",
     benchmark:
-        Path(benchmark_dir, 'bcftools_bcf2vcf_delly_' + config["analysis"]["case_id"] + ".tsv")
+        benchmark_dir + 'delly_cnv_tumor_only_' + config["analysis"]["case_id"] + ".tsv"
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:104]:
```suggestion
        benchmark_dir + 'delly_cnv_tumor_normal_' + config["analysis"]["case_id"] + ".tsv"
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 7]
```diff
@@ -209,6 +233,101 @@ rule ascat_tumor_normal_merge_output:
...
+
+cp {params.tmpdir}/{params.normal}_cov.bed {output.cov_normal_tiddit};
+
+tabix -p vcf -f {output.vcf};
+
+echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap};
+
+rm -rf {params.tmpdir};
+
+#python TIDDIT.py --sv -p 6 --p_ratio 0.10 --bam tumor.bam -o tumor --ref reference.fasta grep -E "#|PASS" tumor.vcf > tumor.pass.vcf
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:288]: Please remove commented lines if not used

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 8]
```diff
@@ -209,6 +233,101 @@ rule ascat_tumor_normal_merge_output:
...
+        tmp_delly_cnv = temp(vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".tmpdellyCNV.vcf.gz"),
+        delly_cnv= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.vcf.gz",
+        ascat_cnv= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.vcf.gz",
+    benchmark:
+        Path(benchmark_dir, 'bcftools_bcf2vcf_delly_' + config["analysis"]["case_id"] + ".tsv")
+    singularity:
+        Path(singularity_image, config["bioinfo_tools"].get("bcftools") + ".sif").as_posix()
+    params:
+        case_name = config["analysis"]["case_id"],
+        process_cnv= get_script_path("process_CNV.py"),
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:309]:
```suggestion
        process_cnv = get_script_path("process_CNV.py"),
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 9]
```diff
@@ -209,6 +233,101 @@ rule ascat_tumor_normal_merge_output:
...
+        get_threads(cluster_config, "bcftools_bcf2vcf_delly")
+    message:
+        ("Processing SVs and CNVs for {params.case_name}")
+    shell:
+        """
+bcftools view  --threads {threads} -f PASS -O z -o {output.delly_sv} {input.delly_sv};
+
+bcftools view --threads {threads} -f PASS -O z -o {output.tmp_delly_cnv} {input.delly_cnv};
+
+python process_cnv -f {output.tmp_delly_cnv} -c delly | bgzip -l 9 -c > {output.delly_cnv};
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:320]:
```suggestion
python {params.process_cnv} -f {output.tmp_delly_cnv} -c delly | bgzip -l 9 -c > {output.delly_cnv};
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule [hunk 10]
```diff
@@ -113,32 +113,74 @@ echo -e \"{params.tumor}\\tTUMOR\" > {output.namemap};
...
+        delly_sv = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".delly.bcf",
+        delly_cnv = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".delly.bcf",
     output:
-        vcf_sv = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".dellysv.vcf.gz",
-        vcf_cnv= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.vcf.gz",
+        delly_sv = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".dellysv.vcf.gz",
+        cnv = temp(vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".tmpdellyCNV.vcf.gz"),
+        delly_cnv= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.vcf.gz",
     benchmark:
         Path(benchmark_dir, 'bcftools_bcf2vcf_delly_' + config["analysis"]["case_id"] + ".tsv")
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule:163]:
```suggestion
        Path(benchmark_dir, 'bcftools_process_SV_CNV_' + config["analysis"]["case_id"] + ".tsv")
```

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule [hunk 11]
```diff
@@ -113,32 +113,74 @@ echo -e \"{params.tumor}\\tTUMOR\" > {output.namemap};
...
         ("Converting BCF from delly to VCF for {params.case_name}")
     shell:
         """
-bcftools view  --threads {threads} -f PASS -O z -o {output.vcf_sv} {input.bcf_sv};
+bcftools view  --threads {threads} -f PASS -O z -o {output.delly_sv} {input.delly_sv};
+
+bcftools view --threads {threads} -f PASS -O z -o {output.cnv} {input.delly_cnv};
 
-bcftools view --threads {threads} -f PASS -O z -o {output.vcf_cnv} {input.bcf_cnv}
+python process_cnv -f {output.cnv} -c delly | bgzip -l 9 -c > {output.delly_cnv};
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule:179]:
```suggestion
python {params.process_cnv} -f {output.cnv} -c delly | bgzip -l 9 -c > {output.delly_cnv};
```

File: BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule (hunk 3 above)
[the remark this answers is not in the document]
Comment (khurrammaqbool) [reply BALSAMIC/snakemake_rules/annotation/vcf2cytosure_convert.rule:45]: I am not sure how do you define "affect" with respect to cgh conversion, but combining the calls from different sources along with curated coverage serves the purpose well at this point.

**Comment by ashwini06** [comment]
@khurrammaqbool : You can also link this issue  https://github.com/Clinical-Genomics/BALSAMIC/issues/605 to the PR description . As a test, I would suggest visualizing the actual output file `.cgh`  in the cytosure interpretation software. You can ask by sending the output file to some who got access to this software (maybe some from the clinic eg: customer or jesper/daniel)

**Comment by khurrammaqbool** [comment]
> @khurrammaqbool : You can also link this issue #605 to the PR description . As a test, I would suggest visualizing the actual output file `.cgh` in the cytosure interpretation software. You can ask by sending the output file to some who got access to this software (maybe some from the clinic eg: customer or jesper/daniel)

It was tested before the implementation.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 12]
```diff
@@ -90,26 +90,50 @@ echo -e \"{params.tumor}\\tTUMOR\\n{params.normal}\\tNORMAL\" > {output.namemap}
...
+        namemap= vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".dellycnv.sample_name_map",
     benchmark:
-        Path(benchmark_dir, 'bcftools_bcf2vcf_delly_' + config["analysis"]["case_id"] + ".tsv")
+        benchmark_dir + 'delly_cnv_tumor_only_' + config["analysis"]["case_id"] + ".tsv"
     singularity:
-        Path(singularity_image, config["bioinfo_tools"].get("bcftools") + ".sif").as_posix()
+        Path(singularity_image, config["bioinfo_tools"].get("delly") + ".sif").as_posix(),
     params:
+        tmpdir = tempfile.mkdtemp(prefix=tmp_dir),
+        housekeeper_id= {"id": config["analysis"]["case_id"],"tags": "clinical"},
```
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:109]: Decided to keep it for now and do the filtering in Hermes @khurrammaqbool

**Pull request merged**